### PR TITLE
:bug: (rules) fix error handling

### LINKED ISSUE #3139: [Bug]: Applying a rule to set all fields to (nothing) results in internal error
### Verified issue does not already exist?

- [X] I have searched and found no existing issue
- [X] I will be providing steps how to reproduce the bug (in most cases this will also mean uploading a demo budget file)

### What happened?

After creating a rule to set all fields to (nothing) and selecting a transaction, I clicked "Apply actions" and an internal error message popped up: 

Untitled2

### Where are you hosting Actual?

Other

### What browsers are you seeing the problem on?

Chrome

### Operating System

Windows 11

### PR DESCRIPTION
**Pull request by MatissJanis**
Closes #3139

Also patching some other modals. They would always close. Even if there were errors.

### CODE REVIEW

File: packages/desktop-client/src/components/modals/EditRule.jsx [hunk 1]
```diff
@@ -879,7 +879,16 @@ export function EditRule({ defaultRule, onSave: originalOnSave }) {
...
+        let usedErrorIdx = 0;
+        setActionSplits(
+          actionSplits.map(item => ({
+            ...item,
+            actions: item.actions.map(action => ({
+              ...action,
+              error: error.actionErrors[usedErrorIdx++] ?? null,
+            })),
+          })),
+        );
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/modals/EditRule.jsx:891]: The addition of splits in rules (+ some other changes) had broken error handling.

So this particular change patches the splits in rules part..

File: packages/desktop-client/src/components/modals/EditRule.jsx [hunk 2]
```diff
@@ -1145,13 +1155,7 @@ export function EditRule({ defaultRule, onSave: originalOnSave }) {
...
                 >
                   <Button onClick={close}>Cancel</Button>
-                  <Button
-                    variant="primary"
-                    onPress={() => {
-                      onSave();
-                      close();
-                    }}
-                  >
+                  <Button variant="primary" onPress={() => onSave(close)}>
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/modals/EditRule.jsx:1158]: We don't want to close the modal if it has errors when saving.

File: packages/loot-core/src/server/accounts/rules.ts [hunk 3]
```diff
@@ -471,6 +471,10 @@ export class Action {
       this.type = 'id';
     }
 
+    if (field === 'account') {
+      assert(value, 'no-null', `Field cannot be empty: ${field}`);
+    }
```
Comment (MatissJanis) [inline packages/loot-core/src/server/accounts/rules.ts:476]: `account` field MUST not be empty. It is a mandatory field.

File: packages/loot-core/src/server/rules/app.ts [hunk 4]
```diff
@@ -15,22 +15,22 @@ function validateRule(rule: Partial<RuleEntity>) {
...
-      .filter((res): res is string => typeof res === 'string');
+        throw e;
+      }
+      return null;
+    });
 
-    return result.length ? result : null;
+    return result.filter((res): res is string => typeof res === 'string').length
+      ? result
+      : null;
```
Comment (MatissJanis) [inline packages/loot-core/src/server/rules/app.ts:33]: There was a bug here. IDK for how long.. The code comment is correct, but the logic is not correct.

Before: returns only the error strings.

After: returns error strings + `null` values. This allows to match the error strings with the faulty actions.

**Review by matt-fidd — APPROVED** [review]
Out of interest, is there a way to force an error for testing purposes in the EditRule modal?

**Comment by MatissJanis** [comment]
> Out of interest, is there a way to force an error for testing purposes in the EditRule modal?

Try creating a "account is (nothing)" rule. That should not work - should give you an error msg.

**Review by matt-fidd — APPROVED** [review]
Looks good!
image

**Pull request merged**